bika/lims/content/abstractroutineanalysis.py

```diff
@@ -442,6 +442,12 @@ def getDependencies(self, retracted=False):
         if not calc:
             return []
 
+        # If the calculation this analysis is bound does not have analysis
+        # keywords (only interims), no need to go further
+        service_uids = calc.getRawDependentServices()
+        if not service_uids:
```

Comment:
I'm not sure how much a type conversion "costs", because the empty list must be converted to a boolean value first. What do you think? Shall we better go for `len(service_uids) == 0`?

Reply:
Hehe, we are going into deep detail here.... We could change `not service_uids` by `len(service_uids)` but don't think that will have a significant impact.

The cost came because `self.getSiblings()` was called, even if this was not necessary because the calculation did not rely on results of other analyses. For you to get an idea, before this change, the function `getDependencies` took ~0.01 sec. to execute for analyses with no dependencies (if their samples have been waken up already, otherwise, the time it takes can easily rise to 1s the first time the function is called). With the change, it takes ~0.0001, no matter if the Sample has been waken up or not. It might seem insignificant, but take into consideration the function is called once per analysis plus one per each sibling (other analyses from same Sample). So if you submit 50 analysis at once, and their Samples have 10 analyses assigned each), the total number of calls is 50*10, that will take about ~5s. Note with this change we don't check siblings, so with this "insignificant" change, the total time reduction in such scenario is 4.9999 s.

Reply:
On another not, I want to revisit and improve performance of `getSiblings` and `getDependents` soon. They can be improved a lot and together with this change, not only the submit will become faster, but also guards